runtime/dart_isolate.cc

```diff
@@ -0,0 +1,685 @@
+                                                  mapping->GetSize()))) {
+    return false;
+  }
+
+  return true;
+}
+
+static bool LoadSnapshot(std::unique_ptr<fml::Mapping> mapping,
+                         bool is_kernel) {
+  if (is_kernel) {
+    return LoadKernelSnapshot(std::move(mapping));
+  } else {
+    return LoadScriptSnapshot(std::move(mapping));
+  }
+  return false;
+}
+
+FXL_WARN_UNUSED_RESULT
+bool DartIsolate::PrepareForRunningFromSnapshot(
+    std::unique_ptr<fml::Mapping> mapping) {
+  TRACE_EVENT0("flutter", "DartIsolate::PrepareForRunningFromSnapshot");
+  if (phase_ != Phase::LibrariesSetup) {
+    return false;
+  }
+
+  if (DartVM::IsRunningPrecompiledCode()) {
+    return false;
+  }
+
+  if (!mapping || mapping->GetSize() == 0) {
+    return false;
+  }
+
+  tonic::DartState::Scope scope(this);
+
+  if (!Dart_IsNull(Dart_RootLibrary())) {
+    return false;
+  }
+
+  if (!LoadSnapshot(std::move(mapping), vm_->GetPlatformKernel() != nullptr)) {
+    return false;
+  }
+
+  if (Dart_IsNull(Dart_RootLibrary())) {
+    return false;
+  }
+
+  if (!MarkIsolateRunnable()) {
+    return false;
+  }
+
+  phase_ = Phase::Ready;
+  return true;
+}
+
+static bool FileNameIsDill(const std::string& name) {
+  const std::string suffix = ".dill";
+
+  if (name.size() < suffix.size()) {
+    return false;
+  }
+
+  if (name.rfind(suffix, name.size()) == name.size() - suffix.size()) {
+    return true;
+  }
+  return false;
+}
+
+bool DartIsolate::PrepareForRunningFromSource(
+    const std::string& main_source_file,
+    const std::string& packages) {
+  TRACE_EVENT0("flutter", "DartIsolate::PrepareForRunningFromSource");
+  if (phase_ != Phase::LibrariesSetup) {
+    return false;
+  }
+
+  if (DartVM::IsRunningPrecompiledCode()) {
+    return false;
+  }
+
+  if (main_source_file.empty()) {
+    return false;
+  }
+
+  tonic::DartState::Scope scope(this);
+
+  if (!Dart_IsNull(Dart_RootLibrary())) {
+    return false;
+  }
+
+  auto& loader = file_loader();
+
+  if (!packages.empty()) {
+    auto packages_absolute_path = files::AbsolutePath(packages);
+    FXL_DLOG(INFO) << "Loading from packages: " << packages_absolute_path;
+    if (!loader.LoadPackagesMap(packages_absolute_path)) {
+      return false;
+    }
+  }
+
+  auto main_source_absolute_path = files::AbsolutePath(main_source_file);
+  FXL_DLOG(INFO) << "Loading from source: " << main_source_absolute_path;
+
+  // The "source" file may be a ".dill" file.
+  if (FileNameIsDill(main_source_absolute_path)) {
+    auto mapping =
+        std::make_unique<fml::FileMapping>(main_source_absolute_path, false);
+    if (mapping == nullptr) {
+      return false;
+    }
+    if (!LoadKernelSnapshot(std::move(mapping))) {
+      return false;
+    }
+  } else {
+    if (tonic::LogIfError(loader.LoadScript(main_source_absolute_path))) {
+      return false;
+    }
+  }
+
+  if (Dart_IsNull(Dart_RootLibrary())) {
```

Comment:
This check is redundant with Dart_IsolateMakeRunnable below.

Reply:
Done.